src/Generator/EnumGenerator/EnumGenerator.php

```diff
@@ -0,0 +1,146 @@
+<?php
+
+namespace Laminas\Code\Generator\EnumGenerator;
+
+use Laminas\Code\Generator\EnumGenerator\Cases\CaseFactory;
+use Laminas\Code\Generator\MethodGenerator;
+
+use function array_key_exists;
+use function array_map;
+use function implode;
+
+final class EnumGenerator
+{
+    /**
+     * Line feed to use in place of EOL
+     */
+    private const LINE_FEED = "\n";
+
+    /**
+     * spaces of indentation by default
+     */
+    private const INDENTATION = '    ';
+
+    private Name $name;
+
+    private CasesInterface $cases;
+
+    /** @var list<MethodGenerator> */
+    private array $methods;
+
+    /**
+     * @param list<MethodGenerator> $methods
+     */
+    private function __construct(Name $name, CasesInterface $cases, array $methods)
+    {
+        $this->name    = $name;
+        $this->cases   = $cases;
+        $this->methods = $methods;
+    }
+
+    public function generate(): string
+    {
+        $output = '';
+
+        if (null !== $this->name->getNamespace()) {
+            $output .= 'namespace ' . $this->name->getNamespace() . ';' . self::LINE_FEED . self::LINE_FEED;
+        }
+
+        $output .= 'enum ' . $this->name->getName() . $this->retrieveType() . ' {'
+            . self::LINE_FEED
+            . $this->retrieveCases()
+            . $this->retrieveMethods()
+            . '}'
+            . self::LINE_FEED;
+
+        return $output;
+    }
+
+    private function retrieveType(): string
+    {
+        if ($this->cases->getBackedType() === null) {
+            return '';
+        }
+
+        return ': ' . ($this->cases->getBackedType() ?? '');
+    }
+
+    private function retrieveCases(): string
+    {
+        return implode(
+            '',
+            array_map(
+                fn (string $case): string => self::INDENTATION . 'case ' . $case . ';' . self::LINE_FEED,
+                $this->cases->getCases()
+            )
+        );
+    }
+
+    /**
+     * @psalm-param array{
+     *      name: string,
+     *      pureCases?: list<string>,
```

Comment:
The declaration of one of either `pureCases` or `backedCases` needing to exist should be ported here too, as shown in the `CaseFactory` suggestions